### DISCUSSION TITLE
Prometheus getting OOM killed

### DISCUSSION THREAD

**Discussion by rsbala** [prometheus/prometheus#9936] [Q&A]
**What did you do?**
Deployed Prometheus 2.10.0 using https://github.com/helm/charts/tree/master/stable/prometheus

**What did you expect to see?**
Prometheus server up and running

**What did you see instead? Under which circumstances?**
Within 5 minutes RAM usage exceeds 80GB and gets OOM killed

**Environment**

* Number of Head series:

	3M

* Number of Head chunk:

	3M

* Prometheus version:

	2.10.0

* Questions

	* Is this RAM usage expected for the above mentioned head series ?
	* One of the labels (source IP) in a metric has high cardinality which is not needed. When I tried to drop the label I get the error ``"Error on ingesting out-of-order samples" num_dropped"``. Is there a way to merge the time series under source_ip label before ingestion so that head series count will not explode. 
	* Any other suggestions most welcome.

I have already removed the unwanted targets and metrics.

**Comment by Nexucis** [comment]
Hello !

Since the version v2.10, Prometheus released many bufixes and optimizations. So it's totally worth to upgrade Prometheus, and that would be my first advise here

> Deployed Prometheus 2.10.0 using https://github.com/helm/charts/tree/master/stable/prometheus

You should probably also take a look at https://github.com/prometheus-community/helm-charts. That's where you have the latest version of the helm chart for Prometheus.

**Comment by rsbala** [comment]
Thanks. As advised I used the new helm chart and Prometheus is upgraded to 2.26.0. Memory usage is reduced a little. It is ~68GB for 2.6M series. Would you please confirm if this is expected ? Any optimization to decrease the memory ? I have attached TSDB status for more details.

![tsdb_status_1](https://user-images.githubusercontent.com/6130541/143271337-adad585b-57c6-49c1-8936-6b7ddc7326e4.png)

![tsdb_status_2](https://user-images.githubusercontent.com/6130541/143271329-a38d824f-aaeb-4685-a81f-2a8059b0729c.png)

**Comment by Nexucis** [comment]
Can you give more details about your Prometheus configuration ? Like how many days do you keep the data ?

In case you are scraping too many timeseries, an idea would be to split the scraping in multiple different Prometheus.
Then you will probably need Thanos to be able to query and aggregate the data across the different Prometheus.

**Comment by rsbala** [comment]
Thanks for the suggestion on Thanos. Currently we have two Prometheus - one for Istio http metrics and another for application metrics. Splitting further might add additional maintenance over ahead. We want to first explore the alternatives without splitting.

```
retention: "3d"
scrape_interval: 30s
scrape_timeout: 10s
evaluation_interval: 1m
```
I am able to remove few more unwanted timeseries and reduce the # Head series from 2.6M to 1.6M. But not much change in the memory usage - 70G (from kubectl top pods).

To further understand, I dropped all the metrics from the scrape targets (in `metric_relabel_configs` -> `keep` only for an unavailable label value). # Head series reduced to 15K (all internal to Prometheus). Screenshots below. Memory usage was still high at 45G. This explains why there is not much reduction in memory usage earlier.

1. We believed memory usage depends mostly on number of head series count. It may not be the case here ? 
2. Isnt the memory usage high for no metrics ingested and just scraping the targets (from ~3000 pods) ? 
3. Can you please help us understand what is going wrong here ? How to optimize ?

Please let me know if any other details needed.

![tsdb_status_new_1](https://user-images.githubusercontent.com/6130541/143453638-9e3e6bc2-20fb-4aac-9211-eb93ce3cd302.png)
![tsdb_status_new_2](https://user-images.githubusercontent.com/6130541/143453657-58ba534f-0a2c-4da9-8a88-a0b6919a2037.png)

**Comment by rsbala** [comment]
More updates. Upgraded to Prometheus v2.31.1. No improvements in Memory usage.

Go profiling using command `go tool pprof -symbolize=remote -inuse_space "http://<prom-server>/debug/pprof/heap"` showed 38G which is only ~40% of the 90G shown in `kubectl top pods`. Major consumption is for `textParse (*PromParser) metrics`: 20G. I guess this is the memory used for the time series ?

I tried to reduce the cache memory but `storage.local.target-heap-size` is deprecated from 2.X.

What could be accounting for the remaining 52G ? Memory leak ? Is there a way to figure out what could be the culprit ? Any help will be much appreciated.

![inuse_latest](https://user-images.githubusercontent.com/6130541/143898587-db89c03e-7799-485a-a088-e945bbade838.png)

**Comment by Nexucis** [comment]
I'm not really an expert around this particular topic unfortunately :/. 

From my point of view, I fear you will have to split the targets across more Prometheus. 

Perhaps @roidelapluie would you have more ideas ?

**Comment by Nexucis** [comment]
Another way around would be to see how much targets you are dropping in the discovery. I think it impacts more the CPU side and not really the memory, but can another way to optimize. 

One day in my company we had an issue with the k8s discovery where we actually asked to Prometheus to get all the pods from the API regardless the different namespace, and then we asked to Prometheus to filter them using relabeling configuration.

It cost a lot of CPUs on our side (and probably memories as well). So a way to avoid it was to list the namespaces Prometheus should use when querying the k8s API.

**Comment by rsbala** [comment]
Thanks for the suggestions. We have already listed the interested namespaces in `kubernetes_sd_configs`. In "/service_discovery" I see 3000/10000 active targets. I guess the dropped 7000 are targets (containers, ports) generated from containers other than istio-proxy and dropped via `relabel_configs`.

`go_memstats_heap_inuse_bytes` (actual memory used) = 45G.
`go_memstats_heap_idle_bytes - go_memstats_heap_released_bytes` (unused memory not yet released to OS) = 25G.
This confirms 70G memory is actually consumed. But pprof inuse_space accounts only for 30G (metric: 17G). I was expecting pprof inuse_space (30G) to match with go_memstats_heap_inuse_bytes (45G). It will be helpful to understand where the missing 15G is accounted for.

**Comment by JessicaGreben** [comment]
There is a calculator here that you can see what consumption should be: https://www.robustperception.io/how-much-ram-does-prometheus-2-x-need-for-cardinality-and-ingestion.

**Comment by SCLogo** [comment]
@rsbala Do you use istio ? If yes, which metrics and targets you removed? We were trying to reduce the labels, but from time to time we are running to OOMkill too
